documentation/docs/explanations/deadlock-analysis.mdx

```diff
@@ -0,0 +1,179 @@
+---
+description: Learn how Stronghold was analysed to prevent any deadlock from occuring
+image: /img/logo/Stronghold_icon.png
+keywords:
+- concurrency
+- deadlock
+- locks
+---
+
+# Deadlock analysis for Stronghold
+
+Current Stronghold use `RwLock` to enable [concurrency](./concurrency.md).
+This approach has the pros of being simple in theory and has fast performance when done right. The downside is that in practice it is usually really difficult to use locks in concurrent programming and you often encounter _data races_ or _deadlocks_ when using them.
+One of the worst case would to have a deadlock in Stronghold which would have the consequences to block Stronghold and not being able to use it.
+To avoid such situation a deadlock analysis has been used which we present here.
+
+## Requirements for multithreaded Strongold
+- Be able to clone the elements parts of the API and use them concurrently. Types reachable through the API that may be cloned and shared between threads 
+  - `Stronghold`
+  - `Client`
+  - `Store`
+- Use `Mutex` or `RWLock`
+- Argument on the impossibility of deadlocks
+
+## Analysis of deadlocks
+This is internal thoughts stemming from experience but does not have peer-reviewed paper backing it.
+
+### Model
+Locks are used to control access to shared memory.
+To simplify the problem we only model mutex so that only a single thread
+can access to a shared memory.
+
+#### Model of function and locks
+Here how we represent a function and locks.
+
+![](./assets/function_model.drawio.png)
+
+- $1$, $2$, $3$ represent locks that protect 3 different and distinct shared memory locations
+- $f_1$ is a function that sequentially takes hold of lock $1$ then $2$ then $3$
+- $f_1$ releases the locks $1$, $2$, $3$ at the end of its computation 
+
+Here we can see that lock $2$ depends on lock $1$ because in $f_1$ you can't take lock $2$ before taking lock $1$. In the same way lock $3$ depends on locks $1$ and $2$.
```

Comment:
```suggestion
### Model

Locks are used for controlling access to shared memory.
To simplify the problem, we only model mutex so that only a single thread can access shared memory.

#### Model of functions and locks

![How functions and locks are represented in Stronghold](./assets/function_model.drawio.png)

- $1$, $2$, $3$ represent locks that protect 3 different and distinct shared memory locations.
- $f_1$ is a function that sequentially takes hold of lock $1$ then $2$ then $3$.
- $f_1$ releases the locks $1$, $2$, $3$ at the end of its computation.

Note that lock $2$ depends on lock $1$, because in $f_1$ you can't take lock $2$ before taking lock $1$. In the same way, lock $3$ depends on locks $1$ and $2$.

```